cpp/dolfinx/fem/FiniteElement.h

```diff
@@ -115,37 +113,80 @@ class FiniteElement
                                           Eigen::Dynamic, Eigen::RowMajor>>& X,
       const Eigen::Tensor<double, 3, Eigen::RowMajor>& J,
       const Eigen::Ref<const Eigen::Array<double, Eigen::Dynamic, 1>>& detJ,
-      const Eigen::Tensor<double, 3, Eigen::RowMajor>& K,
-      const std::uint32_t permutation_info) const;
+      const Eigen::Tensor<double, 3, Eigen::RowMajor>& K) const;
 
-  /// Tabulate the reference coordinates of all dofs on an element
-  /// @return The coordinates of all dofs on the reference cell
-  ///
-  /// @note Throws an exception if dofs cannot be associated with points
-  const Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>&
-  dof_reference_coordinates() const;
-
-  /// Map values of field from physical to reference space which has
-  /// been evaluated at points given by dof_reference_coordinates()
-  void transform_values(
-      ufc_scalar_t* reference_values,
-      const Eigen::Ref<const Eigen::Array<ufc_scalar_t, Eigen::Dynamic,
-                                          Eigen::Dynamic, Eigen::RowMajor>>&
-          physical_values,
-      const Eigen::Ref<const Eigen::Array<double, Eigen::Dynamic,
-                                          Eigen::Dynamic, Eigen::RowMajor>>&
-          coordinate_dofs) const;
-
-  /// Return the number of sub elements (for a mixed element)
-  int num_sub_elements() const;
+  /// Get the number of sub elements (for a mixed element)
+  /// @return the Number of sub elements
+  int num_sub_elements() const noexcept;
 
   /// Return simple hash of the signature string
-  std::size_t hash() const;
+  std::size_t hash() const noexcept;
 
   /// Extract sub finite element for component
   std::shared_ptr<const FiniteElement>
   extract_sub_element(const std::vector<int>& component) const;
 
+  /// Check if interpolation into the finite element space is an
+  /// identity operation given the evaluation on an expression at
+  /// specific points, i.e. the degree-of-freedom are equal to point
+  /// evaluations. The function will return `true` for Lagrange
+  /// elements.
+  ///  @return True is interpolation is an identity operation
+  bool interpolation_ident() const noexcept;
+
+  /// Points on the reference cell at which an expression need to be
+  /// evaluated in order to interpolate the expression in the finite
+  /// element space. For Lagrange elements the points will just be the
+  /// nodal positions. For other elements the points will typically be
+  /// the quadrature points used to evaluate moment degrees of freedom.
+  /// @return Points on the reference cell. Shape is (num_points, tdim).
+  Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
+  interpolation_points() const noexcept;
```

Comment:
What about Hermite elements? You'd need values of derivatives at some points.

Reply:
And we support Hermite elements?

This is already a big step forward from 'dof_coordinates', so we don't need to complicate for cases that aren't yet supported. It can be easily generalised later for other cases. There are several ways this could be done.

Reply:
I guess in future basix could return an int telling which many derivatives are needed at each point. But we can do this in a future PR